python/ray/air/execution/resources/placement_group.py

```diff
@@ -0,0 +1,216 @@
+    ) -> List[Union[RemoteRayObject]]:
+        # If we have an empty head, we schedule the first object (the "head") in the
+        # first PG bundle with num_cpus=0. The second object will also be scheduled in
+        # the first bundle, but will occupy the respective resources. Thus, we start
+        # counting from i = -1, and bundle = max(i, 0) which will be [0, 0, 1, 2, ...]
+        if self.resource_request.head_bundle_is_empty:
+            start = -1
+            bundles = [{}] + self.resource_request.bundles
+        else:
+            start = 0
+            bundles = self.resource_request.bundles
+
+        if len(objects) > len(bundles):
+            raise RuntimeError(
+                f"The number of objects to annotate ({len(objects)}) cannot "
+                f"exceed the number of available bundles ({len(bundles)})."
+            )
+
+        annotated = []
+        for i, (obj, bundle) in enumerate(zip(objects, bundles), start=start):
+            bundle = bundle.copy()
+            num_cpus = bundle.pop("CPU", 0)
+            num_gpus = bundle.pop("GPU", 0)
+            memory = bundle.pop("memory", 0.0)
+
+            annotated.append(
+                obj.options(
+                    scheduling_strategy=PlacementGroupSchedulingStrategy(
+                        placement_group=self.placement_group,
+                        # Max ensures that empty head bundles are correctly placed
+                        placement_group_bundle_index=max(0, i),
+                        placement_group_capture_child_tasks=True,
+                    ),
+                    num_cpus=num_cpus,
+                    num_gpus=num_gpus,
+                    memory=memory,
+                    resources=bundle,
+                )
+            )
+        return annotated
+
+
+@DeveloperAPI
+class PlacementGroupResourceManager(ResourceManager):
+    """Resource manager using placement groups as the resource backend.
+
+    Ray core does not emit events when resources are available. Instead, we
+    have to periodically request and update the scheduling state of the
+    placement groups.
+
+    Internally, the placement group lifecycle is like this:
+    - Resources are requested with ``request_resources()``
+    - A placement group is scheduled ("staged")
+    - A ``PlacementGroup.ready()`` is scheduled ("staging future")
+    - We update the scheduling state when we need to
+        (e.g. when ``has_resources_ready()`` is called)
+    - When staging futures resolved, a placement group is moved from "staging"
+        to "ready"
+    - When a resource request is canceled, we remove a placement group from "staging".
+        If there are not staged placement groups (because they are already "ready"),
+        we remove one from "ready" instead.
+    - When a resource is acquired, the pg is removed from "ready" and moved
+        to "acquired"
+    - When a resource is freed, the pg is removed from "acquired" and destroyed
+
```

Comment:
I found this section confusing with the introduction of the concept of "staging" on top of the existing ResourceManager/PlacementGroup concepts. Maybe we can unify on "requested"?

Also some of the wording (e.g. "removed from XYZ") is currently describing the internal state of the ResourceManager rather than the ResourceRequest.

Reply:
I've moved the description of the internal flow into a code comment in the init function and imoproved the wording of the docstring. This was internal implementation requested by a previous round of reviews